xds/src/test/java/io/grpc/xds/internal/certprovider/MeshCaCertificateProviderTest.java

```diff
@@ -0,0 +1,538 @@
+      Metadata.Key.of(AuthHttpConstants.AUTHORIZATION, Metadata.ASCII_STRING_MARSHALLER);
+  private static final String ZONE = "us-west2-a";
+  private static final long START_DELAY = 200_000_000L;  // 0.2 seconds
+  private static final long[] DELAY_VALUES = {START_DELAY, START_DELAY * 2, START_DELAY * 4};
+  private static final long RPC_TIMEOUT_MILLIS = 100L;
+  /**
+   * Expire time of cert SERVER_0_PEM_FILE.
+   */
+  private static final long CERT0_EXPIRY_TIME_MILLIS = 1899853658000L;
+  /**
+   * Cert validity of 12 hours for the above cert.
+   */
+  private static final long CERT0_VALIDITY_MILLIS = TimeUnit.MILLISECONDS
+      .convert(12, TimeUnit.HOURS);
+  /**
+   * Compute current time based on cert expiry and cert validity.
+   */
+  private static final long CURRENT_TIME_NANOS =
+      TimeUnit.MILLISECONDS.toNanos(CERT0_EXPIRY_TIME_MILLIS - CERT0_VALIDITY_MILLIS);
+  @Rule
+  public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+
+  private static class ResponseToSend {
+    Throwable getThrowable() {
+      throw new UnsupportedOperationException("Called on " + getClass().getCanonicalName());
+    }
+
+    List<String> getList() {
+      throw new UnsupportedOperationException("Called on " + getClass().getCanonicalName());
+    }
+  }
+
+  private static class ResponseThrowable extends ResponseToSend {
+    final Throwable throwableToSend;
+
+    ResponseThrowable(Throwable throwable) {
+      throwableToSend = throwable;
+    }
+
+    @Override
+    Throwable getThrowable() {
+      return throwableToSend;
+    }
+  }
+
+  private static class ResponseList extends ResponseToSend {
+    final List<String> listToSend;
+
+    ResponseList(List<String> list) {
+      listToSend = list;
+    }
+
+    @Override
+    List<String> getList() {
+      return listToSend;
+    }
+  }
+
+  private final Queue<Meshca.MeshCertificateRequest> receivedRequests = new ArrayDeque<>();
+  private final Queue<String> receivedStsCreds = new ArrayDeque<>();
+  private final Queue<String> receivedZoneValues = new ArrayDeque<>();
+  private final Queue<ResponseToSend> responsesToSend = new ArrayDeque<>();
+  private final Queue<String> oauth2Tokens = new ArrayDeque<>();
+  private final AtomicBoolean callEnded = new AtomicBoolean(true);
+
+  @Mock private MeshCertificateServiceGrpc.MeshCertificateServiceImplBase mockedMeshCaService;
+  @Mock private CertificateProvider.Watcher mockWatcher;
+  @Mock private BackoffPolicy.Provider backoffPolicyProvider;
+  @Mock private BackoffPolicy backoffPolicy;
+  @Spy private GoogleCredentials oauth2Creds;
+  @Mock private ScheduledExecutorService timeService;
+  @Mock private TimeProvider timeProvider;
+
+  private ManagedChannel channel;
+  private MeshCaCertificateProvider provider;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    when(backoffPolicyProvider.get()).thenReturn(backoffPolicy);
```

Comment:
nit/optional: you can implement real ones instead of mocks. Might be more readable. But I'm also okay with as-is.

Reply:
I personally like mocks because it's concise and defined "inline" (close to where it is called/used) esp for `BackoffPolicy` which is not the main thing being tested here.